### ISSUE TITLE
Gulp tasks and file read/write profiling

### ISSUE THREAD

**Issue by AntouanK** [gulpjs/gulp#464]
I was wondering whether there is a possibility to have some sort of profiling of our gulp build process.
I discussed about this with @contra last week at the codefront conference, but I still cannot say I have a concrete way of doing this.

For example something like the dev tools 'network' tab, where you can see a waterfall with a timeline and exactly when each file is requested/downloaded/parsed, etc.
![screen shot 2014-05-15 at 10 05 16](https://cloud.githubusercontent.com/assets/4569111/2982094/16e0b78e-dc10-11e3-8877-99da946ee4c1.png)
The visual helps see that stuff happen in parallel.

Similarly we could monitor when a file is read/written from/to the disk, when a task starts/ends, and maybe when each piping/transformation happens, and how long it takes.
I'm new to streams so I'm still trying to figure out if that is even practically possible or not.
But it would help a lot to have a visual way of when and what exactly happens.

First, it would help 'newbies' like me to understand when everything is happening, because with streams it's easy to get lost and to crate bottlenecks.
And secondly, as the project gets bigger, and more files and transformations come in, that would be an easy way to spot bottlenecks or improve performance. I think people working in big projects would find it a big help.

I tried to do some sample monitoring by myself, by I end up 'monkeypatching' functions like fs.readFile to have timestamps and monitor everything.
Doesn't sound like a good idea. But then again, there is no easy way to monitor stuff without tampering with them.
I don't think a plugin can handle something like this, at least not without being invasive.

Would like to hear your ideas/suggestions on this.
I'll try to do some samples and maybe upload something when I have time.

**Comment by stephanos** [comment]
I was just wondering: are there any events Gulp exposes in order to achieve this? For example something like 'started task x', 'finished task x'?

**Comment by stephanos** [comment]
Okay, so I found `logEvents` in events.js that listens on the `start`, `error` and `stop` events. But these only contain duration data for each task. How would one go about timing the inner details of each task?

**Comment by AntouanK** [comment]
@stephanos 
Well, that's the problem. Every "plug-in" must provide details like file read/write timestamps in order to get usable metrics.
Otherwise you'll just get start/end times like gulp already does.

See my comment above https://github.com/gulpjs/gulp/issues/464#issuecomment-44791590
I did something similar by monkey-patching `fs` functions but that's ugly. No other nice way to do it though, unless modules provide that data back to you.
Still, that `fs` read/write timestamps were very useful in order to see when the system does what.

**Comment by contra** [comment]
@AntouanK Why would plugins provide file read/write timestamps? Files don't read or write, vinyl-fs does. Instrument vinyl-fs to get the data you need

**Comment by AntouanK** [comment]
@contra Well, you're right. `gulp.src` and `gulp.dest` read/write files, you could get info from there.
( although you're not sure that the user is not using `fs` or something else )
That's not what I meant though.
I meant that you cannot see what's happening between each `pipe()` in the pipeline.

Let's take a dumb example
```
gulp.src('src/**/*.foo')    // we read 100 files
.pipe( lint() )             // we lint only 50 of them
.pipe( minify() )           // we minify all of them
.pipe( concat() )           // concatenate them
.pipe( gulp.dest('/dest') ) // write 1 file
```

( I tend to not make long tasks, I prefer having small ones and compose them. But most people do make complex to insanely-complex tasks with many pipes in them )

So all we are going to see from the above is reading 100 files and X ms later, a file is written.
Basically what gulp already logs with start/end times of tasks.
No idea about what's happening inside every `pipe()` and how files are transformed/used.
How much time did linting each file take? How about minifying? etc.
I guess an idea might be monkey-patching pipe somehow to track where files go and how long they stay in each stage. But I'm not that good with streams...

**Comment by contra** [comment]
@AntouanK Override the vinyl module and instrument the file objects

**Comment by AntouanK** [comment]
@contra Might try that. Thanks.
We still don't know what module is executed on every pipe though... to log who does what.

**Comment by contra** [comment]
@AntouanK Using the stacktrace from abusing the Error class you can get the path of the file with the code that modified the vinyl file and from that you can extrapolate the module name

**Comment by AntouanK** [comment]
@contra Wanna give a code example?

I can give it a try again once I find time.

Do you think that this "profiling" feature should be a module/plug-in like gulp-util?
Or integrate it in gulp and turn it on with a flag or something?

**Comment by contra** [comment]
@AntouanK external module for sure

**Comment by phated** [comment]
Going to close this, as it will end up in userland.  gulp 3 and 4 both expose lifecycle hooks for tasks (although they have different names) but you will need to use node APIs to hook into streams.  I believe they are working on an API that allows you to profile all async things in the runtime, so I'd look into that.

Closing this.
